### DISCUSSION TITLE
Jinja2 syntax for Netbox

### DISCUSSION THREAD

**Discussion by akanaykin** [netbox-community/netbox#6528] [Q&A] [answered]
H! I want to create my Custom Links but ran into difficulty. I cannot find information on how the objects are correctly named in the Netbox.

For example, my version does not work:
`{% if device.interface.name == 'iLO'%} iLO {% endif%}`

Can you please tell me, did I enter the name of the object incorrectly?
And where can I find more information on Jinja2 for Netbox?

**Comment by candlerb** [comment, marked as answer]
Firstly, the selected device is supplied as "obj", not "device".  see: https://netbox.readthedocs.io/en/stable/additional-features/custom-links/#context-data

For attributes of devices, you need to look at the models in the source code.  For devices, the models are split over several files [here](https://github.com/netbox-community/netbox/tree/develop/netbox/dcim/models).  You can also explore the models interactively in the [netbox shell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/).

Device does not have an attribute called "interface", but it does have one called "interfaces" - because a device can have more than one interface.  Therefore you need to iterate over the interfaces to find the one that you want.  The same applies for IP addresses on an interface - each interface can have multiple IP addresses.

Here is a working example, which iterates over interfaces to find the one with "Management Only" flag set.  You can modify it to match on the name instead if you like (by changing `mgmt_only=True` to `name="iLO"`), although the "Management Only" flag is specifically intended for identifying the ILO or out-of-band management interface without regard to its name.

By making the link text conditional, the button is hidden unless the device has exactly one interface with Management Only set, and that interface has exactly one IP address.

Link text:
```
{% if obj.interfaces.filter(mgmt_only=True).count() == 1 and obj.interfaces.get(mgmt_only=True).ip_addresses.count() == 1 %}Open IPMI{% endif %}
```

Link URL:
```
https://{{ obj.interfaces.get(mgmt_only=True).ip_addresses.first().address.ip }}/
```

**Reply by akanaykin** [reply]
Ohh, it's really good example, thank you! I already use it in my work. And I looked at the material you sent, thanks.

Another question, please tell me please.

I would like to create a Custom Link for Cisco ASA equipment. But clicking on this link should redirect us not to the web page, but to the ASDM application. Do you think it is possible to implement the following idea?

It may also be possible with Custom Links to launch third-party rdp, ssh applications and etc...

**Reply by jeremystretch** [reply]
> But clicking on this link should redirect us not to the web page, but to the ASDM application.

All you can really do is specify an alternate scheme for the URL (for example, `ssh://192.0.2.123`). However, it is entirely up to the user's browser how to handle such links; NetBox cannot control this.

**Reply by akanaykin** [reply]
Thanks, I did it using registry changes.

I'll leave it here, maybe someone will come in handy

```
Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\Software\Classes\asdm]
"URL Protocol"="\"\""
@="\"URL:asdm\""
[HKEY_CURRENT_USER\Software\Classes\asdm\shell]
[HKEY_CURRENT_USER\Software\Classes\asdm\shell\open]
[HKEY_CURRENT_USER\Software\Classes\asdm\shell\open\command]
@="\"C:\\Program Files\\Cisco Systems\\ASDM\\asdm-launcher.jar\" \"%1\""
```

True, none of the browsers was able to launch the asdm-launcher.jar. The browser offered a window asking: "Do you really want to start asdm-launcher.jar"? I confirm, but nothing happens, the program does not start.

Instead of linking to the asdm-launcher.jar:
` @="\"C:\\Program Files\\Cisco Systems\\ASDM\\asdm-launcher.jar\" \"%1\""`,

I indicated a link to the bat file, which I placed in the same directory:
`@="\"C:\\Program Files\\Cisco Systems\\ASDM\\asdm.bat\" \"%1\""`

asdm.bat file contents:
```
nircmd win hide title "%ComSpec%"
@echo off
cd "C:\Program Files (x86)\Cisco Systems\ASDM"
start /b java -jar asdm-launcher.jar
exit 
```
Only with this scheme did it turn out to run asdm.
